### ISSUE TITLE
What is the potential problem on the implementation of CustomMSF?

### ISSUE THREAD

**Issue by yang-jin-hai** [alafage/deep-traffic-generation#1]
Hi,

I'm searching for pytorch implementation of reparameterization of GMM distribution, and then find your awesome work.

In the CustomMSF class, you comment that the it might be incorrect.

I'm wondering that what is the potential problem?

In my case I will use a GMM whose components are all 1-D gaussians. The sampling can be done by discretely sample a component and then sample from the chosen component.

To make it learnable, I think we can apply Gumbel-Softmax to simulate the discrete sampling and then us rsample on the chosen component, and that's what you have done.

So, I'm wondering what problem can be here? Does it work well on your project?

Thanks in advance.

**Comment by alafage** [comment]
Hi,

Thank you for reaching out.

If I'm remembering well, there might be some issues with the gradients and it didn't work on my project. I used several components and it was exactly the same as using only one gaussian. So I'm pretty sure there's something wrong.

Moreover, Tensorflow has an implementation way more complex so it seems to me that it can't be that simplified. I just found this repository (I didn't test it): https://github.com/vsimkus/torch-reparametrised-mixture-distribution. It might be what you're looking for ;)
